windows: fix warnings of MinGW -Wall -O3

starwing opened the pull request:
sweep MinGW warnings.

still two warnings remains:

```
src/win/fs.c: In function 'fs__utime_handle':
src/win/fs.c:1187:3: warning: dereferencing type-punned pointer will break stric
t-aliasing rules [-Wstrict-aliasing]
   TIME_T_TO_FILETIME((time_t) atime, &filetime_a);
   ^
src/win/fs.c:1188:3: warning: dereferencing type-punned pointer will break stric
t-aliasing rules [-Wstrict-aliasing]
   TIME_T_TO_FILETIME((time_t) mtime, &filetime_m);
   ^
```

warnings fixed:
- unsigned/signed unmatched, use int i to iteration array with ARRAY_SIZE(...) -- which is unsigned
- uv__crt_invalid_parameter_handler not used on MinGW
- several uninitialized warning for several places. but some are caused by assert(0); and abort(); (but MinGW doesn't know it will abort)
- some Windows API returns int, not DWORD.
- InterlockedExchange's first argument is volatile LONG_, not unsigned long_.

starwing committed:
windows: fix warnings of MinGW -Wall -O3

indutny commented on src/win/pipe.c:
```diff
@@ -1040,7 +1040,7 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
   int result;
   uv_tcp_t* tcp_send_handle;
-  uv_write_t* ipc_header_req;
+  uv_write_t* ipc_header_req = NULL; /* fix MinGW uninitialized warning */
```
Please use two spaces betwen colon and comment. `NULL;  /*`

indutny commented on src/win/process-stdio.c:
```diff
@@ -436,6 +437,7 @@ int uv__stdio_create(uv_loop_t* loop,
 
       default:
         assert(0);
+        return -1; /* avoid err not initialzed warning from MinGW */
```
Ditto.

indutny commented on src/win/signal.c:
```diff
@@ -78,7 +78,7 @@ int uv__signal_dispatch(int signum) {
        handle != NULL && handle->signum == signum;
        handle = RB_NEXT(uv_signal_tree_s, &uv__signal_tree, handle)) {
-    unsigned long previous = InterlockedExchange(&handle->pending_signum, signum);
+    unsigned long previous = InterlockedExchange((volatile LONG*)&handle->pending_signum, signum);
```
Does it fit into 80 column limit?

indutny commented on src/win/signal.c:
```diff
@@ -307,12 +307,12 @@ int uv_signal_start(uv_signal_t* handle, uv_signal_cb signal_cb, int signum) {
   assert(req->type == UV_SIGNAL_REQ);
 
-  dispatched_signum = InterlockedExchange(&handle->pending_signum, 0);
+  dispatched_signum = InterlockedExchange((volatile LONG*)&handle->pending_signum, 0);
```
Space between cast and expression `(volatile LONG*) &handle`

indutny commented on src/win/tty.c:
```diff
@@ -698,7 +698,7 @@ void uv_process_tty_read_raw_req(uv_loop_t* loop, uv_tty_t* handle,
 
         /* If the buffer is full, emit it */
-        if (buf_used == buf.len) {
+        if ((size_t)buf_used == buf.len) {
```
ditto.

indutny commented on src/win/process-stdio.c:
```diff
@@ -423,9 +424,9 @@ int uv__stdio_create(uv_loop_t* loop,
 
         /* Make an inheritable copy of the handle. */
-        if (uv__duplicate_handle(loop,
+        if ((err = uv__duplicate_handle(loop,
```
Mind splitting into two statements: one assignment and one `if`?

indutny commented on src/win/poll.c:
```diff
@@ -281,7 +281,7 @@ static SOCKET uv__fast_poll_get_peer_socket(uv_loop_t* loop,
 
   index = -1;
-  for (i = 0; i < ARRAY_SIZE(uv_msafd_provider_ids); i++) {
+  for (i = 0; (size_t)i < ARRAY_SIZE(uv_msafd_provider_ids); i++) {
```
`(size_t) i`

indutny commented:
Generally, LGTM, but has some style issues. @saghul mind reviewing?

starwing commented:
@indutny rebased and force pushed, reviewing again? thank you very much!

indutny commented on src/win/poll.c:
```diff
@@ -145,8 +145,8 @@ static int uv__fast_poll_cancel_poll_req(uv_loop_t* loop, uv_poll_t* handle) {
-  unsigned char mask_events;
-  AFD_POLL_INFO* afd_poll_info;
+  unsigned char mask_events = 0;
+  AFD_POLL_INFO* afd_poll_info = NULL;
```
No point in initializing it here, it is set below in both cases.

starwing replied:
it doesn't initialize on assert(0);/abort(); branch, which we know will abort execute, but gcc doesn't know :(

indutny commented on src/win/poll.c:
```diff
@@ -145,8 +145,8 @@ static int uv__fast_poll_cancel_poll_req(uv_loop_t* loop, uv_poll_t* handle) {
     uv_req_t* req) {
-  unsigned char mask_events;
-  AFD_POLL_INFO* afd_poll_info;
+  unsigned char mask_events = 0;
```
Same thing.

indutny commented on src/win/pipe.c:
```diff
@@ -1040,7 +1040,7 @@ static int uv_pipe_write_impl(uv_loop_t* loop,
   int result;
   uv_tcp_t* tcp_send_handle;
-  uv_write_t* ipc_header_req;
+  uv_write_t* ipc_header_req = NULL;  /* fix MinGW uninitialized warning */
```
Could you please initialize it after all declarations? I think C89 prohibits initializing assignments.

starwing replied:
C89 can initialize assignment, but must be const expression, and NULL is really a const expression.

indutny commented on src/win/poll.c:
```diff
@@ -407,7 +407,7 @@ static void uv__slow_poll_submit_poll_req(uv_loop_t* loop, uv_poll_t* handle) {
 static void uv__slow_poll_process_poll_req(uv_loop_t* loop, uv_poll_t* handle,
     uv_req_t* req) {
-  unsigned char mask_events;
+  unsigned char mask_events = 0;
```
No need in initialization.

starwing replied:
maybe we should add a return; after assert(0); or abort();? or use a gcc no_return attribute (which will add a new function)

indutny commented on src/win/process-stdio.c:
```diff
@@ -327,7 +327,7 @@ int uv__stdio_create(uv_loop_t* loop,
         /* the child. */
         uv_pipe_t* parent_pipe = (uv_pipe_t*) fdopt.data.stream;
-        HANDLE child_pipe;
+        HANDLE child_pipe = INVALID_HANDLE_VALUE;
```
No need in initialization.

indutny commented on src/win/process-stdio.c:
```diff
@@ -266,7 +266,7 @@ int uv__stdio_create(uv_loop_t* loop,
   BYTE* buffer;
   int count, i;
-  int err;
+  int err = 0;
```
Are you sure this is needed? Most likely - no.

indutny commented on src/win/process-stdio.c:
```diff
@@ -423,11 +423,10 @@ int uv__stdio_create(uv_loop_t* loop,
-                             stream_handle,
-                             &child_handle) < 0) {
+        err = uv__duplicate_handle(loop, stream_handle, &child_handle);
+
```
Superfluos whitespace.

indutny commented:
Looks cool, though, I dislike initializing all stuff. Those `assert(0)`s are already a pretty fatal error case, perhaps they should be made an `abort()` s intead.

starwing commented:
okay, I will try using abort(); maybe this can avoid warnings.

starwing commented:
add __declspec(noreturn) and use abort(), clear all warnings without -O3, but still some warnings at -O3 level, how to solve these warnings?

```
gcc -O3 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/win/fs.o src/win/fs.c
src/win/fs.c: In function 'fs__utime_handle':
src/win/fs.c:1187:3: warning: dereferencing type-punned pointer will break stric
t-aliasing rules [-Wstrict-aliasing]
   TIME_T_TO_FILETIME((time_t) atime, &filetime_a);
   ^
src/win/fs.c:1188:3: warning: dereferencing type-punned pointer will break stric
t-aliasing rules [-Wstrict-aliasing]
   TIME_T_TO_FILETIME((time_t) mtime, &filetime_m);
   ^
gcc -O3 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/win/pipe.o src/win/pipe.c
src/win/pipe.c: In function 'uv_pipe_write_impl':
src/win/pipe.c:1226:20: warning: 'ipc_header_req' may be used uninitialized in t
his function [-Wmaybe-uninitialized]
         CloseHandle(ipc_header_req->overlapped.hEvent);
                    ^
gcc -O3 -Wall -Wextra -Wno-unused-parameter -Iinclude -Isrc -Isrc/win -DWIN32_LE
AN_AND_MEAN -D_WIN32_WINNT=0x0600 -c -o src/win/process-stdio.o src/win/process-
stdio.c
src/win/process-stdio.c: In function 'uv__stdio_create':
src/win/process-stdio.c:345:39: warning: 'child_pipe' may be used uninitialized
in this function [-Wmaybe-uninitialized]
         CHILD_STDIO_HANDLE(buffer, i) = child_pipe;
                                       ^
```

starwing commented:
There are still many assert(0) in codes, shall I change all of them to abort() for consistency?

starwing commented:
```
src/win/pipe.c:1226:20: warning: 'ipc_header_req' may be used uninitialized in t
his function [-Wmaybe-uninitialized]
         CloseHandle(ipc_header_req->overlapped.hEvent);
                    ^
```

this maybe a bug -- ipc_header_req will only initialized if handle->ipc, so maybe in `handle->flags & UV_HANDLE_BLOCKING_WRITES` branch ipc_header_req will not initialized, can you review this? @indutny

starwing commented:
fix remain 3 warnings, now there are only one warnings, about `ipc_header_req`, please review :)

saghul commented:
Still want me to review this @indutny?

saghul commented on src/win/error.c:
```diff
@@ -33,6 +33,9 @@
  * Display an error message and abort the event loop.
  */
+#if defined(_MSC_VER) || defined(__MINGW32__)
+    __declspec(noreturn)
```
Since this is windows code, what compile would not need this? Can't we just do it always?

starwing replied:
okay... you are total right :)

saghul commented on src/win/process-stdio.c:
```diff
@@ -140,7 +140,8 @@ static int uv__create_stdio_pipe_pair(uv_loop_t* loop,
   if (child_pipe == INVALID_HANDLE_VALUE) {
     err = GetLastError();
-    goto error;
+    if (err)
```
Why the if? We do know there was an error.

starwing replied:
because gcc doesn't know err is always non-zero, so it warnings about not initialized child_pipe, has any ideas about this?

saghul replied:
Uh, but child_pipe is actually initialized to INVALID_HANDLE. If the compiler is stupid, I'd rather leave the warning in, than have weird code I'll scratch my head about later. FTR, no warning is printed on VS.

starwing replied:
but the variable gcc complain _is_ child_pipe_ptr, i.e. it complains the caller's child_pipe. we just doesn't initialize child_pipe_ptr if a error occurs.
so, add a child_pipe_ptr assignment on failure path?

saghul replied:
You can assign it to INVALID_HANDLE on top, just like child_pipe IMHO.

saghul commented:
I'm not sure about replacing assert(0) with abort(). I'm generally more inclined to fail gracefully.

starwing commented:
removed `#ifdef`s.

starwing commented:
okay:) fix this, now this patch resolve all -Wall -O3 warnings from gcc, except this one:

```
src/win/pipe.c:1226:20: warning: 'ipc_header_req' may be used uninitialized in t
his function [-Wmaybe-uninitialized]
         CloseHandle(ipc_header_req->overlapped.hEvent);
                    ^
```

hope for review this warning, and tell us whether its a bug or how to fix this, thanks! @indutny @saghul

saghul commented:
I'll run the test suite on VS a bit later, to check that we didn't create warning there :-)

@piscisaureus, what's your stand on the assert/abort thing?

starwing commented:
if you use VS2013, you can open /sdl check, sdl stands for software development lifetime, it may give you some warnings about initializing. I tried last time and notice several warnings.

starwing commented:
I have tested on VS2013, no bug report, I will open /sdl to see further

starwing commented:
`ipc_header_req` is nearly a bug -- /sdl check also warning about that. /sdl check also report fs.c(789), tty(216), but I have checked these, they are safe.

starwing commented:
hi dear @saghul has some news on this topic?

saghul commented:
LGTM for the most part, but I still don't like the `abort()`s. I'd like a second opinion from @indutny or @piscisaureus on that.

starwing commented:
@saghul I don't like abort(), though. what about add a return; after assert(0)? notice that assert will disable in Relese mode, a return may insure the control flow is correct.
Another way is use uv_fatal_error, to fail gracefully in these case.
how some opinions?

starwing commented:
@saghul Hi, Sorry for the long time repy, I'm busy these days.

I have removed all of `abort()`, using `assert(0) + return;` instead. still hoping for a LGTM :)

several improvement: 
- removed `abort()`, as mentioned above. 
- fix another warning from pipe.c, introduced by previous commits.
- tested on VS2013 and run tests (5 test failure, unfortunately).
- fix all warnings include the `ipc_header_req` ones (add a assert for safe).

hopes for a review and merge, thank you :)

saghul commented:
@starwing thnaks! I'll have it one last look tonight.

saghul commented:
Thanks @starwing! Landed in https://github.com/joyent/libuv/commit/1ec4c234388c0548db149772b2b8fd510a6dd843.
